core/src/main/java/org/bitcoinj/core/Context.java

```diff
@@ -1,16 +1,119 @@
+     * object. This method returns that. Note that to help you develop, this method will <i>also</i> propagate whichever
+     * context was created last onto the current thread, if it's missing. However it will print an error when doing so
+     * because propagation of contexts is meant to be done manually: this is so two libraries or subsystems that
+     * independently use bitcoinj (or possibly alt coin forks of it) can operate correctly.
+     *
+     * @throws java.lang.IllegalStateException if no context exists at all.
+     */
+    public static Context get() {
+        Context tls = slot.get();
+        if (tls == null) {
+            if (lastConstructed == null)
+                throw new IllegalStateException("You must construct a Context object before using bitcoinj!");
+            slot.set(lastConstructed);
+            log.error("Performing thread fixup: you are accessing bitcoinj via a thread that has not had any context set on it.");
+            log.error("This error has been corrected for, but doing this makes your app less robust.");
+            log.error("You should use Context.propagate() or a ContextPropagatingThreadFactory.");
+            log.error("Please refer to the user guide for more information about this.");
+            // TODO: Actually write the user guide section about this.
+            // TODO: If the above TODO makes it past the 0.13 release, kick Mike and tell him he sucks.
+            return lastConstructed;
+        } else {
+            return tls;
+        }
+    }
+
+    // A temporary internal shim designed to help us migrate internally in a way that doesn't wreck source compatibility.
+    public static Context getOrCreate(NetworkParameters params) {
+        Context context;
+        try {
+            context = get();
+        } catch (IllegalStateException e) {
+            log.warn("Implicitly creating context. This is a migration step and this message will eventually go away.");
+            context = new Context(params);
+            return context;
+        }
+        if (context.getParams() != params)
+            throw new IllegalStateException("Context does not match implicit network params: " + context.getParams() + " vs " + params);
+        return context;
+    }
 
-    protected Context() {
-        confidenceTable = new TxConfidenceTable();
+    /**
+     * Sets the given context as the current thread context. You should use this if you create your own threads that
+     * want to create core BitcoinJ objects. Generally, if a class can accept a Context in its constructor and might
+     * be used (even indirectly) by a thread, you will want to call this first. Your task may be simplified by using
+     * a {@link org.bitcoinj.utils.ContextPropagatingThreadFactory}.
+     *
+     * @throws java.lang.IllegalStateException if this thread already has a context
+     */
+    public static void propagate(Context context) {
+        checkState(slot.get() == null);
```

Comment:
Might be helpful to use the form of `checkState()` that includes a message.  Something like

    checkState(..., "Thread %s already has a Context for %s", Thread.currentThread(), slot.get().getParams().getId());